### ISSUE TITLE
How to throw specific exceptions for certain rules after the removal of OnFailure/OnAnyFailure?

### ISSUE THREAD

**Issue by AnysSido** [FluentValidation/FluentValidation#1842]
### Is your feature request related to a problem? Please describe.

I am using FluentValidation in an MVC application and currently handle validation errors in two different ways:

Unexceptional/expected errors such as invalid user input are wrapped in a result object and bound to ModelState by my controllers so they can be displayed to the user. They are written like this:

`RuleFor(v => v.Name).NotEmpty().WithMessage("Role name cannot be empty")`

Exceptional/unexpected errors such as an invalid entity ID on an edit request throw an exception. They are currently written like this:

`RuleFor(v => v.RoleId).GreaterThan(0).OnFailure(x => throw new MyException(nameof(x.RoleId)));`

Different errors are thrown depending on the issue, for example I throw a RecordNotFoundException when an edit is attempted for an entity ID that does not exist.

How can I achieve this same behavior with the removal of OnFailure/OnAnyFailure?

### Describe the solution you'd like

An alternative way to achieve the same result as above.

### Additional Context

Opened as a feature request as suggested [here](https://github.com/FluentValidation/FluentValidation/issues/1725#issuecomment-896700375)

**Comment by JeremySkinner** [comment]
Hi, throwing exceptions part way through the validation process isn't recommended. The usual way of handling this would be to associate the exception as custom metadata with a particular rule using `WithState`. You can then inspect the failures collection and act upon this metadata as necessary. Quick proof of concept:

```csharp
class MyValidator : AbstractValidator<Foo> {
  MyValidator() {
    RuleFor(v => v.RoleId).GreaterThan(0).WithState(x => new MyException(nameof(x.RoleId)));
  }
}

var validator = new MyValidator();
var result = validator.Validate(new Foo());

foreach (var failure in result.Errors) {
  if (failure.CustomState is Exception ex) {
    throw ex;
  }
}
```

**Comment by AnysSido** [comment]
Awesome thank you, that works well. One issue with this approach, though it's not the end of the world due to exceptions being rare, is that if we hit an exceptional rule, the rest of the rules will still run which is a bit of a waste if they're executing against the db.

For example:
```c#
RuleFor(v => v.RoleId)
        .GreaterThan(0).WithException(cmd => new ArgumentException(nameof(cmd.RoleId)))
        .MustAsync(Exist).WithException(cmd => new RecordNotFoundException());

RuleFor(v => v.Name)
        .NotEmpty().WithMessage("Role name cannot be empty")
        .MustAsync(BeUnique).WithMessage(cmd => $"A role with the name {cmd.Name} already exists");
```

Is it possible for me to somehow short-circuit the validation inside my WithException extension method? WithException is just a type-constrained wrapper around WithState.

**Comment by JeremySkinner** [comment]
Mhm not really. What you can do is set the cascade mode to stop at the validator level, this will prevent the validator from continuing execution when the first rule fails:

```csharp
public class MyValidator() : AbstractValidator<Foo> {
  public MyValidator() {
    CascadeMode = CascadeMode.Stop;
    // ... rules ...
  }
}
```

...but this affects all rules not just those that throw exceptions.
